docs/langref/relay_pattern.rst

```diff
@@ -0,0 +1,143 @@
+        diamond = is_op('add')(path1, path2)
+
+        # Expr
+        inp = relay.var('input')
+        weight = relay.var('weight')
+        conv2d = relay.op.nn.conv2d(inp, weight)
+        relu = relay.op.nn.relu(conv2d)
+        leaky_relu = relay.op.nn.leaky_relu(conv2d, alpha=0)
+        out = relu + leaky_relu
+
+        # Check
+        assert diamond.match(out)
+
+The final example we would like to match which is not yet implemented in the prototype is matching diamonds with a post-dominator relationship. Our plan is to embed dominator analysis as type of matching in the pattern language in order to allow for pattern matching with unknown topology.  This is important because we want to able to use the language to describe fuse patterns, like elementwise operations followed by a conv2d::
+
+    def test_match_dom_diamond():
+        # Pattern
+        is_conv2d = is_op('nn.conv2d')(is_input(), is_input())
+        reduction = is_op('add')(wildcard(), wildcard())
+    	diamond = dominates(is_conv2d, is_elemwise, reduction)
+
+        # Expr
+        inp = relay.var('input')
+        weight = relay.var('weight')
+        conv2d = relay.op.nn.conv2d(inp, weight)
+        relu = relay.op.nn.relu(conv2d)
+        leaky_relu = relay.op.nn.leaky_relu(conv2d, alpha=0)
+        out = relu + leaky_relu
+
+        # Check
+        assert diamond.match(out)
+
+Design
+======
+
+The pattern language proposed is designed to be a mirror of Relay's IR with additional support for common scenarios. The goal of the pattern language is to provide a regular-expression like capability for matching data-flow graphs and doing rewriting.
+
+The high level design is to introduce a language of patterns for now we propose the language as::
+
+    Pattern ::= expr
+            | *
+            | pattern(pattern1, ... patternN)
+            | has_type(pattern, type)
+            | has_attr(pattern, attr, attr_value)
+            | is_input(name)
+            | pattern1 `|` pattern2
+            | dominates(parent_pattern, path_pattern, child_pattern)
+
+The above language then provides a matching interface with both can select sub-graphs as well as verify that the graph does match the pattern.
+
+Expression Pattern
+******************
+
+Match a literal expression.
+
+Wildcard
+******************
+
+Match any expression.
+
+Type Pattern
+******************
+
+Check that the expression matched by the nested pattern has a particular type.
+
+Attribute Pattern
+******************
+
+Check that the operator matched by the pattern has an attribute with a particular value.
+
+Input
+******************
+
+Check that the expression is an input, i.e has no parents and is a variable.
+
+
+Alternate
+******************
+
+Either match the first pattern or the second pattern.
+
+Domination
+******************
+
+Match the parent pattern for the route node and then check that the child pattern holds for each child along the domination path.
```

Comment:
@mbrookhart In the doc, Can you give an example or explanation of what the pattern for path should be like? In the test, you have

```
    is_unary_elemwise = (wildcard().has_attr("TOpPattern", K_ELEMWISE))(wildcard())
```

I guess it is supposed to match arbitrary number of elemwise ops, but it is not clear to me how it works, given that there is no pattern to describe "repeating" (kleene star). What is the role of second wildcard?

Reply:
ok I read the code, I think this is more like a pattern to describe a node in a path, rather than a path itself. `child pattern` in L143 should be `path pattern`, I think.

Reply:
Rewrote the sentence, it didn't make much sense, thanks!